File: othello_csharp/Othello/ColorPrint.cs
//==========================================================
// Colorprint
// Interface for colored printing
// Akseli Lukkarila
// 2019-2023
//==========================================================

using Pastel;
using System;
using System.Drawing;

namespace Othello
{
    public static class ColorPrint
    {
        public static string Get<T>(T text, Color color)
        {
            return $"{text}".Pastel(color);
        }

        public static void Write<T>(T text, Color color)
        {
            Console.Write($"{text}".Pastel(color));
        }

        public static void WriteLine<T>(T text, Color color)
        {
            Console.WriteLine($"{text}".Pastel(color));
        }

        /// Print error message with red colour.
        public static void Error(string message)
        {
            var (indent, text) = SplitLeadingWhitespace(message);
            Console.WriteLine($"{indent}Error: {message}".Pastel(Color.Red));
        }

        /// Print warning message with yellow colour.
        public static void Warn(string message)
        {
            var (indent, text) = SplitLeadingWhitespace(message);
            Console.WriteLine($"{indent}Warning: {text}".Pastel(Color.Yellow));
        }

        /// Split a string to the leading whitespace and rest of the string.
        private static (string, string) SplitLeadingWhitespace(string message)
        {
            // Find the index of the first non-whitespace character.
            int indentSize = 0;
            foreach (char c in message)
            {
                if (char.IsWhiteSpace(c))
                    indentSize++;
                else
                    break;
            }

            return (message[..indentSize], message[indentSize..]);
        }
    }
}

File: othello_swift/Sources/othello_swift/colorprint.swift
//==========================================================
// Colorprint
// Interface for colored printing
// Akseli Lukkarila
// 2019-2023
//==========================================================

import ColorizeSwift

/// Format string with colour.
func getColor<T>(_ message: T, _ color: TerminalColor) -> String {
    "\(message)".foregroundColor(color)
}

/// Print text with specified colour.
func printColor<T>(_ message: T, _ color: TerminalColor) {
    print(getColor(message, color))
}

/// Print error message with red colour.
func printError(_ message: String) {
    let (indent, text) = splitLeadingWhitespace(message)
    print("\(indent)Error: \(text)".red())
}

/// Print warning message with yellow colour.
func printWarn(_ message: String) {
    let (indent, text) = splitLeadingWhitespace(message)
    print("\(indent)Warning: \(text)".yellow())
}

/// Split a string into the leading whitespace and the rest of the string.
func splitLeadingWhitespace(_ message: String) -> (String, String) {
    let indentSize = message.prefix(while: { $0.isWhitespace }).count
    let index = message.index(message.startIndex, offsetBy: indentSize)
    return (String(message[..<index]), String(message[index...]))
}
